src/frontend/src/optimizer/plan_node/generic/vector_index_lookup_join.rs

```diff
@@ -0,0 +1,147 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use pretty_xmlish::Pretty;
+use risingwave_common::catalog::{ColumnDesc, Field, Schema};
+use risingwave_common::types::{DataType, StructType};
+use risingwave_pb::common::PbDistanceType;
+use risingwave_pb::plan_common::PbVectorIndexReaderDesc;
+
+use crate::OptimizerContextRef;
+use crate::catalog::TableId;
+use crate::expr::{ExprDisplay, ExprImpl, InputRef};
+use crate::optimizer::plan_node::generic::{GenericPlanNode, GenericPlanRef};
+use crate::optimizer::property::FunctionalDependencySet;
+
+#[derive(Debug, Clone, educe::Educe)]
+#[educe(Hash, PartialEq, Eq)]
+pub struct VectorIndexLookupJoin<PlanRef> {
+    pub input: PlanRef,
+    pub top_n: u64,
+    pub distance_type: PbDistanceType,
+    pub index_name: String,
+    pub index_table_id: TableId,
+    pub info_column_desc: Vec<ColumnDesc>,
+    pub info_output_indices: Vec<usize>,
+    pub include_distance: bool,
+
+    pub vector_column_idx: usize,
+    pub hnsw_ef_search: Option<usize>,
+    #[educe(Hash(ignore), Eq(ignore))]
+    pub ctx: OptimizerContextRef,
+}
+
+impl<PlanRef: GenericPlanRef> GenericPlanNode for VectorIndexLookupJoin<PlanRef> {
+    fn functional_dependency(&self) -> FunctionalDependencySet {
+        // TODO: copy the one of input, and extend with an extra column with no dependency
+        FunctionalDependencySet::new(
+            self.info_output_indices.len() + if self.include_distance { 1 } else { 0 },
+        )
+    }
+
+    fn schema(&self) -> Schema {
+        let mut schema = self.input.schema().clone();
+        schema.fields.push(Field::new(
+            "vector_info",
+            DataType::list(
+                StructType::new(
+                    self.info_output_indices
+                        .iter()
+                        .map(|idx| {
+                            (
+                                self.info_column_desc[*idx].name.clone(),
+                                self.info_column_desc[*idx].data_type.clone(),
+                            )
+                        })
+                        .chain(
+                            self.include_distance
+                                .then(|| [("__distance".to_owned(), DataType::Float64)].into_iter())
+                                .into_iter()
+                                .flatten(),
+                        ),
+                )
+                .into(),
+            ),
+        ));
+        schema
+    }
+
+    fn stream_key(&self) -> Option<Vec<usize>> {
+        self.input.stream_key().map(|key| key.to_vec())
```

Comment:
Just to highlight this function is filled up by this refactor PR. It was returning None.